### PR TITLE
[SPARK-13430][PySpark][ML] Python API for training summaries of linear and logistic regression

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

Adding Python API for training summaries of LogisticRegression and LinearRegression in PySpark ML.
## How was this patch tested?

Added unit tests to exercise the api calls for the summary classes.  Also, manually verified values are expected and match those from Scala directly.

### CODE REVIEW

File: python/pyspark/mllib/common.py [hunk 1]
```diff
@@ -130,20 +130,39 @@ def callMLlibFunc(name, *args):
...
-        self._sc._gateway.detach(self._java_model)
+        self._sc._gateway.detach(self._java_obj)
+
+    @classmethod
+    def fromActiveSparkContext(cls, java_obj):
+        """Create from a currently active context"""
+        sc = SparkContext._active_spark_context
+        return cls(sc, java_obj)
 
     def call(self, name, *a):
```
Comment (BryanCutler) [inline python/pyspark/mllib/common.py]: this ends up in the api docs because ML has the `:inherited-members:` flag on [here](https://raw.githubusercontent.com/apache/spark/master/python/docs/pyspark.ml.rst)

Is there any way to avoid this besides making these private?

File: python/pyspark/ml/wrapper.py [hunk 2]
```diff
@@ -34,10 +34,12 @@ class JavaWrapper(Params):
 
     __metaclass__ = ABCMeta
 
-    #: The wrapped Java companion object. Subclasses should initialize
-    #: it properly. The param values in the Java object should be
-    #: synced with the Python wrapper in fit/transform/evaluate/copy.
-    _java_obj = None
```
Comment (BryanCutler) [inline python/pyspark/ml/wrapper.py]: @mengxr @jkbradley it seems like this should not be a static member, right?  From what I could tell, it's not even being used since the assignment is always to `self._java_obj`.  So I made a change to make it a member instead.

Comment (BryanCutler) [inline python/pyspark/ml/wrapper.py]: Nevermind, I opened a separate JIRA [SPARK-13937](https://issues.apache.org/jira/browse/SPARK-13937) for this since it is not really related to the issue in this PR.

File: python/pyspark/ml/classification.py [hunk 3]
```diff
@@ -231,6 +232,210 @@ def intercept(self):
...
+    @property
+    @since("2.0.0")
+    def hasSummary(self):
+        """
+        Indicates whether a training summary exists for this model
+        instance.
+        """
+        return self._call_java("hasSummary")
+
+    """
```
Comment (thunterdb) [inline python/pyspark/ml/classification.py]: if the code should not be committed, there is no need to include it in comments

File: python/pyspark/ml/classification.py [hunk 4]
```diff
@@ -231,6 +232,210 @@ def intercept(self):
...
+    Abstraction for Logistic Regression Results for a given model.
+
+    .. versionadded:: 2.0.0
+    """
+
+    @property
+    @since("2.0.0")
+    def predictions(self):
+        """
+        Dataframe outputted by the model's `transform` method.
```
Comment (thunterdb) [inline python/pyspark/ml/classification.py]: nit: technically, `outputted` is not in the dictionary.

File: python/pyspark/ml/regression.py [hunk 5]
```diff
@@ -151,6 +151,209 @@ def intercept(self):
...
+    @property
+    @since("2.0.0")
+    def hasSummary(self):
+        """
+        Indicates whether a training summary exists for this model
+        instance.
+        """
+        return self._call_java("hasSummary")
+
+    """
```
Comment (thunterdb) [inline python/pyspark/ml/regression.py]: same thing here

File: python/pyspark/ml/wrapper.py [hunk 6]
```diff
@@ -223,3 +223,20 @@ def _call_java(self, name, *args):
         sc = SparkContext._active_spark_context
         java_args = [_py2java(sc, arg) for arg in args]
         return _java2py(sc, m(*java_args))
+
+
+class JavaCallable(object):
```
Comment (thunterdb) [inline python/pyspark/ml/wrapper.py]: this looks good to me, but @mengxr will have a more expert opinion on this class.

**Comment by thunterdb** [comment]
@BryanCutler  thanks for the PR, I only have a few small comments

File: python/pyspark/ml/classification.py (hunk 3 above)
Comment (BryanCutler) [inline python/pyspark/ml/classification.py]: Sure I can remove this if that's preferable.  I just didn't want to ignore it without knowing the status of the Scala API.  @mengxr @jkbradley is `evaluate` going to public soon or is it best to not include this at all?

File: python/pyspark/ml/wrapper.py (hunk 6 above)
Comment (BryanCutler) [inline python/pyspark/ml/wrapper.py]: Ideally, I would have liked to reuse parts from `JavaWrapper` and `JavaModel` but just making this small class was the most simple solution and least intrusive.

**Comment by BryanCutler** [comment]
Thanks for checking this out and the comments @thunterdb!

File: python/pyspark/ml/classification.py (hunk 3 above)
Comment (jkbradley) [inline python/pyspark/ml/classification.py]: I want to make it public.  I'll send a PR now.

File: python/pyspark/ml/wrapper.py (hunk 6 above)
Comment (jkbradley) [inline python/pyspark/ml/wrapper.py]: JavaCallable seems reasonable.  Could you modify it so that JavaModel can inherit from it and eliminate the duplicate code?

Comment (BryanCutler) [inline python/pyspark/ml/wrapper.py]: I tried that before and it got a little messy with the definition of `_java_obj`, but I think I have an idea so let me give it another shot.

**Comment by BryanCutler** [comment]
Made `JavaModel` inherit `JavaCallable`, please take a look when you get a chance @jkbradley and see if this change is ok, thanks!

File: python/pyspark/ml/wrapper.py [hunk 7]
```diff
@@ -190,8 +190,30 @@ def _transform(self, dataset):
         return DataFrame(self._java_obj.transform(dataset._jdf), dataset.sql_ctx)
 
 
+class JavaCallable(object):
```
Comment (yanboliang) [inline python/pyspark/ml/wrapper.py]: Can we implement this function by updating ```JavaWrapper```?

Comment (BryanCutler) [inline python/pyspark/ml/wrapper.py]: `JavaWrapper` does define a java object wrapper variable, but also does more since it is derived from `Params`.  Since I need to wrap a Java Summary object (which is just a plain Java object), it doesn't really make sense to use it unless we want to change that.

Actually, I think it might make more sense to have `JavaWrapper` inherit from `JavaCallable` and only define `_java_obj` there.  Then rename the classes a little like this to better reflect what they are for:

`JavaCallable(object)` -> `JavaWrapper(object)`
and
`JavaWrapper(Params)` -> `JavaParamWrapper(JavaWrapper, Params)`

That might be better done in another JIRA though, because of the name change

Comment (jkbradley) [inline python/pyspark/ml/wrapper.py]: I agree with having JavaWrapper inherit from JavaCallable.  Renaming sounds reasonable, but doing it in another PR sounds like a good idea.

File: python/pyspark/ml/classification.py [hunk 8]
```diff
@@ -28,6 +28,7 @@
 
 
 __all__ = ['LogisticRegression', 'LogisticRegressionModel',
+           'BinaryLogisticRegressionSummary', 'BinaryLogisticRegressionTrainingSummary',
```
Comment (jkbradley) [inline python/pyspark/ml/classification.py]: LogisticRegressionSummary and LogisticRegressionTrainingSummary should probably be public too

File: python/pyspark/ml/classification.py [hunk 9]
```diff
@@ -233,6 +234,210 @@ def intercept(self):
         """
         return self._call_java("intercept")
 
+    @property
+    @since("2.0.0")
+    def summary(self):
+        """
+        Gets summary (e.g. residuals, mse, r-squared ) of model on
+        training set. An exception is thrown if
+        `trainingSummary == None`.
```
Comment (jkbradley) [inline python/pyspark/ml/classification.py]: trainingSummary is None (for Python)

File: python/pyspark/ml/classification.py [hunk 10]
```diff
@@ -233,6 +234,210 @@ def intercept(self):
...
+        """
+        java_blrt_summary = self._call_java("summary")
```
Comment (jkbradley) [inline python/pyspark/ml/classification.py]: Add note that this will need to be updated once multiclass support is added.

File: python/pyspark/ml/classification.py [hunk 11]
```diff
@@ -233,6 +234,210 @@ def intercept(self):
...
+
+    @since("2.0.0")
+    def evaluate(self, dataset):
+        """
+        Evaluates the model on a test dataset.
+
+        :param dataset:
+          Test dataset to evaluate model on.
+        """
+        java_blr_summary = self._call_java("evaluate", dataset)
```
Comment (jkbradley) [inline python/pyspark/ml/classification.py]: Check that this is a DataFrame

File: python/pyspark/ml/classification.py [hunk 12]
```diff
@@ -233,6 +234,210 @@ def intercept(self):
...
+        Dataframe outputted by the model's `transform` method.
+        """
+        return self._call_java("predictions")
+
+    @property
+    @since("2.0.0")
+    def probabilityCol(self):
+        """
+        Field in "predictions" which gives the calibrated probability
+        of each instance as a vector.
```
Comment (jkbradley) [inline python/pyspark/ml/classification.py]: "instance" --> "class"  This looks like an error from the Scala doc; could you please fix it there too?

File: python/pyspark/ml/classification.py [hunk 13]
```diff
@@ -233,6 +234,210 @@ def intercept(self):
...
+        """
+        return self._call_java("probabilityCol")
+
+    @property
+    @since("2.0.0")
+    def labelCol(self):
+        """
+        Field in "predictions" which gives the true label of each
+        instance.
```
Comment (jkbradley) [inline python/pyspark/ml/classification.py]: Add: "(if available)"  Same for Scala doc

File: python/pyspark/ml/classification.py [hunk 14]
```diff
@@ -233,6 +234,210 @@ def intercept(self):
...
+    @since("2.0.0")
+    def featuresCol(self):
+        """
+        Field in "predictions" which gives the features of each instance
+        as a vector.
+        """
+        return self._call_java("featuresCol")
+
+
+class LogisticRegressionTrainingSummary(LogisticRegressionSummary):
```
Comment (jkbradley) [inline python/pyspark/ml/classification.py]: ```@inherit_doc```

File: python/pyspark/ml/classification.py [hunk 15]
```diff
@@ -233,6 +234,210 @@ def intercept(self):
...
+    @property
+    @since("2.0.0")
+    def totalIterations(self):
+        """
+        Number of training iterations until termination.
+        """
+        return self._call_java("totalIterations")
+
+
+class BinaryLogisticRegressionSummary(LogisticRegressionSummary):
```
Comment (jkbradley) [inline python/pyspark/ml/classification.py]: ```@inherit_doc```

File: python/pyspark/ml/regression.py [hunk 16]
```diff
@@ -151,6 +151,209 @@ def intercept(self):
         """
         return self._call_java("intercept")
 
+    @property
+    @since("2.0.0")
+    def summary(self):
+        """
+        Gets summary (e.g. residuals, mse, r-squared ) of model on
+        training set. An exception is thrown if
+        `trainingSummary == None`.
```
Comment (jkbradley) [inline python/pyspark/ml/regression.py]: "==" --> "is"

File: python/pyspark/ml/regression.py [hunk 17]
```diff
@@ -151,6 +151,209 @@ def intercept(self):
...
+        the label at each instance.
+        """
+        return self._call_java("predictionCol")
+
+    @property
+    @since("2.0.0")
+    def labelCol(self):
+        """
+        Field in "predictions" which gives the true label of each
+        instance.
```
Comment (jkbradley) [inline python/pyspark/ml/regression.py]: add "(if available)"
Same for Scala doc

File: python/pyspark/ml/regression.py [hunk 18]
```diff
@@ -151,6 +151,209 @@ def intercept(self):
...
+    @property
+    @since("2.0.0")
+    def pValues(self):
+        """
+        Two-sided p-value of estimated coefficients and intercept.
+        """
+        return self._call_java("pValues")
+
+
+class LinearRegressionTrainingSummary(LinearRegressionSummary):
```
Comment (jkbradley) [inline python/pyspark/ml/regression.py]: inherit_doc

Also, add featuresCol and objectiveHistory

File: python/pyspark/ml/tests.py [hunk 19]
```diff
@@ -238,6 +238,17 @@ def setParams(self, seed=None):
         return self._set(**kwargs)
 
 
+class HasThrowableProperty(Params):
```
Comment (jkbradley) [inline python/pyspark/ml/tests.py]: Why move this?

**Comment by jkbradley** [comment]
I had several minor comments, and there were 2 missing methods.  Otherwise, it's looking good.

File: python/pyspark/ml/regression.py [hunk 20]
```diff
@@ -151,6 +151,209 @@ def intercept(self):
...
+    """
+    .. note:: Experimental
+
+    Linear regression training results. Currently, the training summary ignores the
+    training coefficients except for the objective trace.
```
Comment (jkbradley) [inline python/pyspark/ml/regression.py]: "coefficients" --> "weights"  This is a bug from the Scala doc; can you fix it there too?

File: python/pyspark/ml/regression.py [hunk 21]
```diff
@@ -151,6 +151,209 @@ def intercept(self):
...
+        """
+        return self._call_java("tValues")
+
+    @property
+    @since("2.0.0")
+    def pValues(self):
+        """
+        Two-sided p-value of estimated coefficients and intercept.
+        """
+        return self._call_java("pValues")
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: Should we clarify that ```coefficientStandardErrors, tValues and pValues``` are only valid when train ```LinearRegressionModel``` by ```normal``` solver? We may also need to update the Scala doc.

File: python/pyspark/ml/regression.py [hunk 22]
```diff
@@ -151,6 +151,209 @@ def intercept(self):
...
+
+    .. versionadded:: 2.0.0
+    """
+
+    @property
+    @since("2.0.0")
+    def totalIterations(self):
+        """
+        Number of training iterations until termination.
+        """
```
Comment (yanboliang) [inline python/pyspark/ml/regression.py]: ```totalIterations``` is only valid when training model by ```l-bfgs``` solver.

File: python/pyspark/ml/wrapper.py (hunk 7 above)
Comment (BryanCutler) [inline python/pyspark/ml/wrapper.py]: Ok, cool.  So is it good enough as-is for this PR, and then I can open another JIRA for the changes mentioned above?

File: python/pyspark/ml/tests.py (hunk 19 above)
Comment (BryanCutler) [inline python/pyspark/ml/tests.py]: I realized I kind of placed this in a weird place in the file previously, so I moved it next to the test case that was using it to be consistent with the other tests.  It's not related to this PR, so would you prefer I move it back?

File: python/pyspark/ml/regression.py [hunk 23]
```diff
@@ -151,6 +151,228 @@ def intercept(self):
...
+
+    Linear regression training results. Currently, the training summary ignores the
+    training weights except for the objective trace.
+
+    .. versionadded:: 2.0.0
+    """
+
+    @property
+    @since("2.0.0")
+    def featuresCol(self):
```
Comment (BryanCutler) [inline python/pyspark/ml/regression.py]: @jkbradley shouldn't `featuresCol` be part of LinearRegressionSummary instead?

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 24]
```diff
@@ -539,6 +539,10 @@ class LinearRegressionTrainingSummary private[regression] (
  * Linear regression results evaluated on a dataset.
  *
  * @param predictions predictions outputted by the model's `transform` method.
+ * @param predictionCol Field in "predictions" which gives the predicted value of the label at
+ *                      each instance.
+ * @param labelCol Field in "predictions" which gives the true label of each instance
+ *                 (if available).
```
Comment (BryanCutler) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: I added "(if available)" to these descriptions, but isn't the `labelCol` for sure going to be defined in any dataset that was trained or evaluated on that will lead to a summary object?

File: python/pyspark/ml/regression.py (hunk 21 above)
Comment (BryanCutler) [inline python/pyspark/ml/regression.py]: My only thought is that when the solver is `auto` (by default), there is a bit more logic to explain when these are valid and maybe it would be better expressed in the user documentation.  What are your thoughts @jkbradley ?

File: python/pyspark/ml/wrapper.py (hunk 7 above)
Comment (jkbradley) [inline python/pyspark/ml/wrapper.py]: Good as is for this PR, I'd say.  Yes, please create a JIRA.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala (hunk 24 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Hmm, good point; "if available" only makes sense for the generic summary classes, not the *TrainingSummary classes.

File: python/pyspark/ml/tests.py (hunk 19 above)
Comment (jkbradley) [inline python/pyspark/ml/tests.py]: No, that's fine.

File: python/pyspark/ml/regression.py (hunk 21 above)
Comment (jkbradley) [inline python/pyspark/ml/regression.py]: +1 for noting that they may not be available depending on the solver.  How about:
```
This value is only available when using the "normal" solver.
.. seealso:: :py:attr:`LinearRegression.solver`
```

File: python/pyspark/ml/regression.py (hunk 23 above)
Comment (jkbradley) [inline python/pyspark/ml/regression.py]: Yes, it should.  Could you please update it in Python and Scala?

**Comment by jkbradley** [comment]
@BryanCutler Thanks for the updates!  They look good.  I just commented on some of the discussions above.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 25]
```diff
@@ -777,10 +777,10 @@ sealed trait LogisticRegressionSummary extends Serializable {
   /** Dataframe outputted by the model's `transform` method. */
   def predictions: DataFrame
 
-  /** Field in "predictions" which gives the calibrated probability of each instance as a vector. */
+  /** Field in "predictions" which gives the calibrated probability of each class as a vector. */
   def probabilityCol: String
 
-  /** Field in "predictions" which gives the true label of each instance. */
+  /** Field in "predictions" which gives the true label of each instance (if available). */
```
Comment (BryanCutler) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: I removed the "(if available)" from all other descriptions but this generic trait because even in the `LinearRegressionSummary` class, for instance, it assumes that a `labelCol` always exists so it can calculate the summary metrics.

**Comment by BryanCutler** [comment]
jenkins retest this please

**Comment by BryanCutler** [comment]
@jkbradley  I incorporated the latest feedback, please take another look when you get a chance, thanks!

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 25 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: You're right, sounds good

**Comment by jkbradley** [comment]
LGTM
Merging with master
Thanks very much!

Could you please create a JIRA for updating JavaWrapper and JavaCallable?

**Comment by BryanCutler** [comment]
Cool, thanks!
> Could you please create a JIRA for updating JavaWrapper and JavaCallable?

Will do

**Comment by jkbradley** [comment]
By the way, what do you think about providing methods to test if data are available?
E.g.,
```
/** Indicates if [[coefficientStandardErrors()]] is available in this summary */
def hasCoefficientStandardErrors: Boolean
```

**Comment by BryanCutler** [comment]
@jkbradley, adding those methods would make it a little more usable I think.  Another option might be to just add a method to return the solver used, and then the user could key off that before calling the other methods.  It's not as clear-cut as an explicit method returning a bool, but in the case that 'auto' was set, it would be useful info to have and maybe help to explain why certain methods, like `tValues`, are not available.

**Comment by BryanCutler** [comment]
So if we included something like the following in an example, it should be clear what's happening

```Scala
val trainingSummary = lrModel.summary
trainingSummary.getSolver() match {
  case "normal" =>
    println(trainingSummary.coefficientStandardErrors.mkString(","))
    ...
  case "l-bfgs" =>
    println(trainingSummary.totalIterations)
    ...
}
```

**Comment by jkbradley** [comment]
Good point, let's expose the solver for now. I made a JIRA: https://issues.apache.org/jira/browse/SPARK-14461
